### DISCUSSION TITLE
GraphQL links not working

### DISCUSSION THREAD

**Discussion by SabbeRubbish** [gatsbyjs/gatsby#30157] [Help] [answered]
Hi guys/girls,

Writing my first gatsby source plugin #superexcited.
All is working fine, I'm getting my data from external service and created nodes in GraphQL:

```javascript
"allSalonizedAppointments": {
      "nodes": [
        {
          "customer_id": "7508867",
          "id": "27201514",
          "customer": null
        }
      ]
    },
    "allSalonizedCustomers": {
      "nodes": [
        {
          "first_name": "John",
          "last_name": "Doe",
          "id": "7508867"
        },
        {
          "first_name": "Jane",
          "last_name": "Doe",
          "id": "7522368"
        }
      ]
    }
  },
```

As you can see I've already added a "customer" field, trying to populate it with salonizedCustomer type. GraphQL already finds all fields, but the link is giving me issues:

```javascript
exports.createSchemaCustomization = ({ actions, schema }) => {
  const { createTypes } = actions;
  createTypes(`
    type salonizedAppointments implements Node {
      customer_id: String!
      customer: salonizedCustomers @link(from: "salonizedCustomers.id" by: "customer_id")
    }
    type salonizedCustomers implements Node {
      id: String!
    }
  `);
};
```
I've already tried using other syntax using the more explicit syntax, but to no avail.
I'm **guessing** some issues may be related to the fact that the actual data has salonizedAppointments.id as a **String** and salonizedCustomers.id as an **int**, but as I've treated both of them in the schema as **String**s I don't see the issue.

Some help would be really appreciated!

**Comment by vladar** [comment, marked as answer]
Try something like this instead:

```js
exports.createSchemaCustomization = ({ actions, schema }) => {
  const { createTypes } = actions;
  createTypes(`
    type salonizedAppointments implements Node {
      customer_id: String!
      customer: salonizedCustomers @link(from: "customer_id")
    }
    type salonizedCustomers implements Node {
      id: String!
    }
  `);
};
```

**Reply by SabbeRubbish** [reply]
Hi vladar,

Thanks for your reply.
Unfortunately, this doesn't work either.
It tried loading the entire plugin in a default starter to eliminate any other plugins wreaking havoc, but I still have the ``null`` customer.
When I don't manipulate the schema, this is the GraphQL data:
```javascript
{
  "data": {
    "allSalonizedCustomers": {
      "nodes": [
        {
          "id": "7508867",
          "first_name": "John",
          "last_name": "Doe"
        },
        {
          "id": "7522368",
          "first_name": "Jane",
          "last_name": "Doe"
        }
      ]
    },
    "allSalonizedAppointments": {
      "nodes": [
        {
          "customer_id": 7508867
        },
        {
          "customer_id": 7508867
        },
        {
          "customer_id": 7508867
        }
      ]
    }
  },
}
```

**Reply by vladar** [reply]
Hm, maybe the issue is in type mismatch indeed: `customer_id` is int vs string for `id`. Can you cast `customer_id` to `string` when creating nodes? If that doesn't help, can you share your project or a [minimal reproduction][reproduction]?

[reproduction]: https://gatsby.dev/reproduction

**Reply by SabbeRubbish** [reply]
Hi vlad, 

Thanks again, but making them all strings didn't help either.
I've made minimal reproduction in which I also don't succeed in making the foreign key relationship:
https://github.com/SabbeRubbish/bug-repro

**Reply by vladar** [reply]
I think you are using the wrong argument (`by` vs `from`). I've changed [this line](https://github.com/SabbeRubbish/bug-repro/blob/155be122609a7dd114a874f9d6fb310aacc270d6/gatsby-node.js#L67) in your repro and it worked:

```diff
exports.createSchemaCustomization = ({ actions, schema }) => {
    const { createTypes } = actions;
    createTypes(`
      type salonizedAppointments implements Node {
        customer_id: String!
-        customer: salonizedCustomers @link(by: "customer_id")
+        customer: salonizedCustomers @link(from: "customer_id")
      }
      type salonizedCustomers implements Node {
        id: String!
      }
    `);
  };

```

**Reply by SabbeRubbish** [reply]
wow vlad, you found it! 
Not entirely my mistake though, as this [documentation article](https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization/#foreign-key-fields) is stating it incorrectly:
```graphql
type MarkdownRemark implements Node {
  frontmatter: Frontmatter
}
type Frontmatter {
  author: AuthorJson @link # default foreign-key relation by `id`
  reviewers: [AuthorJson] @link(by: "email") # foreign-key relation by custom field
}
type AuthorJson implements Node {
  posts: [MarkdownRemark] @link(by: "frontmatter.author.email", from: "email") # easy back-ref
}
```
Line 6 has the incorrect "by" mentioned that I took over in my example. I'll make a PR to correct that #30174.

But your suggestion to convert all **int**s to **string*s was also important. Both pieces fixed my puzzle!
Thanks!

**Reply by vladar** [reply]
I think documentation is correct 😅 but I do agree that argument naming is confusing.

The argument `from` defines a field on `this` type to use for matching. The argument `by` defines a field on the referenced type to use for matching.

So take this example:

```graphql
type Frontmatter {
  reviewers: [AuthorJson] @link(by: "email") # foreign-key relation by custom field
}
```

It implies that your markdown looks like this:
```md
---
reviewers:
  - jane@example.com
  - zoe@example.com
---
```

And your author JSON looks like this:
```json
[
  {
    "name": "Doe",
    "firstName": "Jane",
    "email": "jane@example.com",
    "joinedAt": "2018-01-01"
  },
  {
    "name": "Doe",
    "firstName": "Zoe",
    "email": "zoe@example.com",
    "receivedSwag": true
  }
]
```

Then you read `@link` directive this way: 

```
Use the value from the field `Frontmatter.reviewers` and match it by the field `AuthorJSON.email`
```
(from is omitted so we use the field value itself as a source for matching).

**Reply by SabbeRubbish** [reply]
Ok that makes sense now :)
Sorry for the bogus PR, I'll retract it (somehow :)).
